keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,1087 @@
+and higher fault tolerance by selecting striping respectively
+mirroring or raid in the storage class parameters of a driver for LVM,
+like for example [TopoLVM](https://github.com/cybozu-go/topolvm).
+
+The maximum size of the resulting volume then depends on the storage
+class and its parameters.
+
+#### Network attached storage
+
+In contrast to local storage, network attached storage can be made
+available on more than just one node. However, for technical reasons
+(high-speed network for data transfer inside a single data center) or
+political reasons (data must only be stored and processed in a single
+jurisdication) availability may still be limited to a subset of the
+nodes in a cluster.
+
+#### Custom schedulers
+
+For situations not handled by the Kubernetes scheduler now and/or in
+the future, a [scheduler
+extender](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/scheduler_extender.md)
+can influence pod scheduling based on the information exposed via the
+new API. The
+[topolvm-scheduler](https://github.com/cybozu-go/topolvm/blob/master/docs/design.md#how-the-scheduler-extension-works)
+currently does that with a driver-specific way of storing capacity
+information.
+
+Alternatively, the [scheduling
+framework](https://kubernetes.io/docs/concepts/configuration/scheduling-framework/)
+can be used to build and run a custom scheduler where the desired
+policy is compiled into the scheduler binary.
+
+#### Operators for applications
+
+Application operators for modern scale out storage services
+(e.g. MongoDB, ElasticSearch, Kafka, MySQL, PostgreSQL, Minio, etc.)
+may want to control creation of volumes carefully in order to optimize
+availability, durability, performance and cost. For more information
+about this, see the [StoragePool API for Advanced Storage Placement
+KEP](https://github.com/kubernetes/enhancements/pull/1347).
+
+The storage pool API as introduced in this KEP enables the creation of such
+operators. When combined with a CSI extension that allows a sidecar to
+list and query storage pools, even more information might become
+available, like health of a storage pool.
+
+## Proposal
+
+### Caching remaining capacity via the API server
+
+CSI defines the `GetCapacity` RPC call for the controller service to
+report information about the current capacity but it's not accessible
+in the Kubernetes scheduler. A new type `CSIStoragePool` type gets
+added to store the result of the `GetCapacity` calls for one storage
+pool in the API server and thus make it available to the scheduler. In
+the future it might get extended to also store other information like
+health of the pool.
+
+### Identifying storage pools
+
+The external-provisioner will be extended to handle the management
+of `CSIStoragePool` objects. For local storage in the node, there will
+be one such `CSIStoragePool` object per node and CSI driver on that
```

Comment:
I think we need to resolve this difference. Currently the CSI spec implies storage pool == storage class per topology. We should either follow this model or fix the csi spec. I don't think we should create more differences between K8s and CSI.

Reply:
I think making this 1-1 mapping would also simplify the API. Right now I'm not sure there's value for multiple storage classes sharing the same storage pool object since capacity is still reported per storage pool.

Reply:
There are several cases where the same storage pool is shared by multiple classes: the common case is to have additional parameters in the storage class, for example `fsType`. That could be ignored, but then there are also opaque parameters where Kubernetes doesn't know whether it can ignore them.

The other case is where the opaque parameters in the storage class are different and tell the CSI driver how to use a pool. The LVM example with stripping vs. mirroring falls into this category.

My understanding is that with simpler API, @msau42 means removing the `CSIStoragePool` and `CSIStorageByClass` types and instead flatten the objects with a `CSIStorageCapacity` type:
```
type CSIStorageCapacity struct {
    NodeTopology v1.NodeSelector
    StorageClassName string
    Capacity resource.Quantity
}
```

Correct?

So instead of one `CSIStoragePool` with multiple `CSIStorageByClass` entries, we would have as many `CSIStorageCapacity` objects as we had `CSIStorageByClass` entries before.

This works for the purpose of this KEP. However, I prefer to keep `CSIStoragePool` for two reasons:
- My expectation is that the `CSIStorageCapacity` objects will put a higher load on the apiserver than `CSIStoragePool`: we have a single writer for each `CSIStoragePool` and my guestimate (sic!) is that the majority of the writes will be triggered by capacity changes that affect all classes, so the entire object will get updated. That's one API call for `CSIStoragePool` and many for `CSIStorageCapacity`. The overall data volume is also lower because we avoid repeating the same information (header, node selector).
- `CSIStoragePool` as a concept and object is needed for https://github.com/kubernetes/enhancements/pull/1347.

Reply:
In the previous comment I showed how "storage pool" to "storage class" is a 1:many relationship.

The opposite case is where there's just a single storage class which selects the CSI driver (no fstype, no parameters) and the driver then manages local storage: that means there are many different pools (at least one per node) and thus a many:1 "storage pool" to "storage class" relationship.

Combine those and we you get many:many.